# Introduction

In this competition, you are provided with 1.5 years of customers behavior data from Santander bank to predict what new products customers will purchase. The data starts at 2015-01-28 and has monthly records of products a customer has, such as "credit card", "savings account", etc. You will predict what additional products a customer will get in the last month, 2016-06-28, in addition to what they already have at 2016-05-28. These products are the columns named: ind_(xyz)_ult1, which are the columns #25 - #48 in the training data. You will predict what a customer will buy in addition to what they already had at 2016-05-28. 

The test and train sets are split by time, and public and private leaderboard sets are split randomly.

# Package Imports

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os
import sys
%matplotlib inline

# Data Loading

About the datasets: \
    train.csv - the training set \
    test.csv - the test set \
    sample_submission.csv - a sample submission file in the correct form

In [2]:
current_dir = os.getcwd()
arr = os.listdir()

In [3]:
PROJ_ROOT = os.path.join(current_dir)

In [4]:
data1 = os.path.join(PROJ_ROOT,
                              "data",
                              "train_ver2.csv")
src_dir = os.path.join(PROJ_ROOT, "src")

data2 = os.path.join(PROJ_ROOT,
                              "data",
                              "Feature Description.xlsx")

In [5]:
# add local python functions
sys.path.append(src_dir)
sys.path.append(data1)
sys.path.append(data2)

In [6]:
#limit_rows = 7000000
load_data = pd.read_csv(data1, dtype={"sexo":str,
                                                    "ind_nuevo":str,
                                                    "ult_fec_cli_1t":str,
                                                    "indext":str})

C:\Users\srafi\AppData\Local\Temp\ipykernel_16272\1236936928.py:2: DtypeWarning: Columns (5,8,11,15) have mixed types. Specify dtype option on import or set low_memory=False.
  load_data = pd.read_csv(data1, dtype={"sexo":str,


In [7]:
df = load_data.copy()

In [8]:
df.shape[0]

13647309

In [9]:
pd.options.display.max_colwidth=None
features = pd.read_excel(data2)


In [10]:
column_maping_dic = {col:features.iloc[i,2] for (col, i)  in zip(features['Column Name'], features.index)}

In [11]:
df.rename(columns=column_maping_dic, inplace=True)

In [12]:
df.columns.tolist()

['date_identifier',
 'customer_code',
 'employee_index',
 'country_resid',
 'sex',
 'age',
 'date_join',
 'customer_registered_in_last6month',
 'customer_seniority_month',
 'customer_loyalty',
 'last_date_asprimary_customer_ifnot_endofmonth',
 'customer_type_atbeginningofmonth',
 'customer_relationtype_atbeginningofmonth',
 'customer_residence_index',
 'customer_foreigner_index',
 'spouse_index',
 'channel_used_tojoin',
 'customer_dceased',
 'address_type',
 'province_code',
 'province_name',
 'activity_index',
 'gross_income_household',
 'customer_segmentation_vip_indiv_colleggrad',
 'saving_account',
 'guarantees',
 'current_accounts',
 'derivada_account',
 'payroll_account',
 'junior_account',
 'más_particular_account',
 'particular_account',
 'particular_plus_account',
 'shortterm_deposits',
 'mediumterm_deposits',
 'longterm_deposits',
 'e_account',
 'funds',
 'mortgage',
 'pensions_plan',
 'loans',
 'taxes',
 'credit_card',
 'securities',
 'home_account',
 'payroll',
 'pensions',

In [13]:
#unique_ids = pd.Series(df['customer_code'].unique())
#limit_people = int(1.2e4)
#unique_id = unique_ids.sample(n=limit_people)
#df = df[df['customer_code'].isin(unique_id)]
df.describe().T

count           mean            std       min  \
customer_code            13647309.0  834904.211501  431565.025784  15889.00   
customer_loyalty         13619575.0       1.178399       4.177469      1.00   
address_type             13619574.0       1.000000       0.000000      1.00   
province_code            13553718.0      26.571470      12.784017      1.00   
activity_index           13619575.0       0.457811       0.498217      0.00   
gross_income_household   10852934.0  134254.318238  230620.238822   1202.73   
saving_account           13647309.0       0.000102       0.010113      0.00   
guarantees               13647309.0       0.000023       0.004812      0.00   
current_accounts         13647309.0       0.655484       0.475210      0.00   
derivada_account         13647309.0       0.000394       0.019844      0.00   
payroll_account          13647309.0       0.080867       0.272631      0.00   
junior_account           13647309.0       0.009474       0.096873      0.00   
más_particular_account   13647309.0       0.009727       0.098143      0.00   
particular_account       13647309.0       0.129008       0.335209      0.00   
particular_plus_account  13647309.0       0.043306       0.203545      0.00   
shortterm_deposits       13647309.0       0.001779       0.042138      0.00   
mediumterm_deposits      13647309.0       0.001661       0.040721      0.00   
longterm_deposits        13647309.0       0.042967       0.202782      0.00   
e_account                13647309.0       0.082744       0.275494      0.00   
funds                    13647309.0       0.018486       0.134701      0.00   
mortgage                 13647309.0       0.005887       0.076498      0.00   
pensions_plan            13647309.0       0.009171       0.095325      0.00   
loans                    13647309.0       0.002627       0.051191      0.00   
taxes                    13647309.0       0.052536       0.223106      0.00   
credit_card              13647309.0       0.044389       0.205957      0.00   
securities               13647309.0       0.025608       0.157962      0.00   
home_account             13647309.0       0.003848       0.061911      0.00   
payroll                  13631246.0       0.054724       0.227441      0.00   
pensions                 13631246.0       0.059429       0.236425      0.00   
direct_debit             13647309.0       0.127916       0.333996      0.00   

                               25%       50%         75%          max  
customer_code            452813.00  931893.0  1199286.00   1553689.00  
customer_loyalty              1.00       1.0        1.00        99.00  
address_type                  1.00       1.0        1.00         1.00  
province_code                15.00      28.0       35.00        52.00  
activity_index                0.00       0.0        1.00         1.00  
gross_income_household    68710.98  101850.0   155955.96  28894395.51  
saving_account                0.00       0.0        0.00         1.00  
guarantees                    0.00       0.0        0.00         1.00  
current_accounts              0.00       1.0        1.00         1.00  
derivada_account              0.00       0.0        0.00         1.00  
payroll_account               0.00       0.0        0.00         1.00  
junior_account                0.00       0.0        0.00         1.00  
más_particular_account        0.00       0.0        0.00         1.00  
particular_account            0.00       0.0        0.00         1.00  
particular_plus_account       0.00       0.0        0.00         1.00  
shortterm_deposits            0.00       0.0        0.00         1.00  
mediumterm_deposits           0.00       0.0        0.00         1.00  
longterm_deposits             0.00       0.0        0.00         1.00  
e_account                     0.00       0.0        0.00         1.00  
funds                         0.00       0.0        0.00         1.00  
mortgage                      0.00       0.0        0.00         1.0

In [14]:
df.shape[0]

13647309

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13647309 entries, 0 to 13647308
Data columns (total 48 columns):
 #   Column                                         Dtype  
---  ------                                         -----  
 0   date_identifier                                object 
 1   customer_code                                  int64  
 2   employee_index                                 object 
 3   country_resid                                  object 
 4   sex                                            object 
 5   age                                            object 
 6   date_join                                      object 
 7   customer_registered_in_last6month              object 
 8   customer_seniority_month                       object 
 9   customer_loyalty                               float64
 10  last_date_asprimary_customer_ifnot_endofmonth  object 
 11  customer_type_atbeginningofmonth               object 
 12  customer_relationtype_atbeginningofmonth

The last month is being used as test data to predict what they buy on top what they have after 2016-05-28. Start of data: 2015-01-28

In [16]:
df['date_identifier'] = pd.to_datetime(df.loc[:,'date_identifier'], format="%Y-%m-%d")
df['date_join'] = pd.to_datetime(df.loc[:,'date_join'], format="%Y-%m-%d")

In [17]:
np.min(df['date_identifier']), np.max(df['date_identifier'])

(Timestamp('2015-01-28 00:00:00'), Timestamp('2016-05-28 00:00:00'))

Creating a month column as the data are on a monthly basis of the product purchased by the customers

In [18]:
df['month'] = df.loc[:,'date_identifier'].dt.month

converting age colum to a numeric column

In [19]:
df.shape[0]

13647309

In [20]:
df['age'] = pd.to_numeric(df['age'], errors='coerce')

In [21]:
missing_percentage = pd.Series(df.isnull().sum() * 100 / df.shape[0])
missing_percentage.sort_values(ascending=False)

spouse_index                                     99.986752
last_date_asprimary_customer_ifnot_endofmonth    99.818330
gross_income_household                           20.475648
customer_segmentation_vip_indiv_colleggrad        1.387585
channel_used_tojoin                               1.363829
customer_type_atbeginningofmonth                  1.097513
customer_relationtype_atbeginningofmonth          1.097513
province_name                                     0.685784
province_code                                     0.685784
sex                                               0.203732
address_type                                      0.203227
customer_dceased                                  0.203220
activity_index                                    0.203220
customer_foreigner_index                          0.203220
customer_residence_index                          0.203220
customer_loyalty                                  0.203220
customer_registered_in_last6month                 0.2032

### Age Distribution 

In [22]:
'''
with sns.plotting_context("notebook",font_scale=1.5):
    sns.set_style("whitegrid")
    ax = sns.histplot(df["age"].dropna(),
                 bins=80,
                 kde=True,
                 color="tomato")
    ax.lines[0].set_color('blue')
    plt.title("Age Distribution")
    plt.ylabel("Count")

'''

'\nwith sns.plotting_context("notebook",font_scale=1.5):\n    sns.set_style("whitegrid")\n    ax = sns.histplot(df["age"].dropna(),\n                 bins=80,\n                 kde=True,\n                 color="tomato")\n    ax.lines[0].set_color(\'blue\')\n    plt.title("Age Distribution")\n    plt.ylabel("Count")\n\n'

As seen in the above figure, there are alot of small and very large ages in the dataset. There are two peaks in the data, one around 20-30 and the other between 40-50. We replace the small ages with average of the ages between 18 and 30 and high ages with means between 30 and 100.

In [23]:
df.loc[df.age<18, 'age'] = df.loc[(df.age>=18) & (df.age<=30), 'age'].mean(skipna=True)
df.loc[df.age>100, 'age'] = df.loc[(df.age>=30) & (df.age<=100), 'age'].mean(skipna=True)
df.age.fillna(df.age.mean(), inplace=True)
df.age = df.age.astype(int)

C:\Users\srafi\AppData\Local\Temp\ipykernel_16272\3540085279.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df.age.fillna(df.age.mean(), inplace=True)


In [24]:
'''
df.loc[df.age<18, 'age'] = df.loc[(df.age>=18) & (df.age<=30), 'age'].mean(skipna=True)
df.loc[df.age>100, 'age'] = df.loc[(df.age>=30) & (df.age<=100), 'age'].mean(skipna=True)
df.age.fillna(df.age.mean(), inplace=True)
df.age = df.age.astype(int)
'''

"\ndf.loc[df.age<18, 'age'] = df.loc[(df.age>=18) & (df.age<=30), 'age'].mean(skipna=True)\ndf.loc[df.age>100, 'age'] = df.loc[(df.age>=30) & (df.age<=100), 'age'].mean(skipna=True)\ndf.age.fillna(df.age.mean(), inplace=True)\ndf.age = df.age.astype(int)\n"

In [25]:
'''
with sns.plotting_context("notebook",font_scale=1.5):
    sns.set_style("whitegrid")
    ax = sns.histplot(df["age"].dropna(),
                 bins=80,
                 kde=True,
                 color="tomato")
    ax.lines[0].set_color('blue')
    plt.title("Age Distribution")
    plt.ylabel("Count")
    '''

'\nwith sns.plotting_context("notebook",font_scale=1.5):\n    sns.set_style("whitegrid")\n    ax = sns.histplot(df["age"].dropna(),\n                 bins=80,\n                 kde=True,\n                 color="tomato")\n    ax.lines[0].set_color(\'blue\')\n    plt.title("Age Distribution")\n    plt.ylabel("Count")\n    '

### New Customer or Not?

In [26]:
df['customer_registered_in_last6month'].value_counts(dropna=False) * 100 / df.shape[0]

customer_registered_in_last6month
 0     93.85270
 1      5.94408
NaN     0.20322
Name: count, dtype: float64

only less than 0.2% of the data are null for registration within last 6 months.

In [27]:
df.customer_registered_in_last6month = pd.to_numeric(df.customer_registered_in_last6month,errors="coerce")

In [28]:
null_registed_in6m = df.loc[df['customer_registered_in_last6month'].isnull(),:].groupby('customer_code', sort=False).size()
null_registed_in6m

customer_code
1050741    2
1051017    2
1051064    4
1051387    3
1048660    5
          ..
618245     1
553680     2
1125581    1
1401509    1
429713     1
Length: 7340, dtype: int64

In [29]:
null_registed_in6m.max()

6

In [30]:
# are they new customers?
df.loc[df['customer_registered_in_last6month'].isnull(),"date_join"].unique()

<DatetimeArray>
['NaT']
Length: 1, dtype: datetime64[ns]

it seems the date of join is null for these customers. How we can know that when these customers joined. It can be found through their first product. If that is within last 6 months, it means that they are new customers. Let's look at their seniority too.

In [31]:
df.loc[df['customer_registered_in_last6month'].isnull(),'customer_seniority_month'].unique()

array(['     NA'], dtype=object)

Hmm, their seniority in months is also null! To be considered as registered within the last 6 months, we considere any customer who has purchased a product on or after 2016-01-28.

In [32]:
mask = df.groupby('customer_code')['date_identifier'].min() <= "2016-01-28"
custcode_buying_last6m = mask.index.to_list()

those customers who have null and have their customer code in the custcode_buying_last6m will receive 1 for the column of customer_registered_in_last6month otherwise, they'll get 0.

In [33]:
df.loc[(df['customer_registered_in_last6month'].isnull() & df['customer_code'].isin(custcode_buying_last6m)),
                     'customer_registered_in_last6month'] = 1

In [34]:
df['customer_registered_in_last6month'].isnull().sum()

0

it seemed that all the null values were for those joined within 6 months! 

In [35]:
df['customer_registered_in_last6month'].unique()

array([0., 1.])

### Customer Seniority in Months

In [36]:
df.customer_seniority_month = pd.to_numeric(df.customer_seniority_month,errors="coerce")
np.sum(df["customer_seniority_month"].isnull())

27734

In [37]:
# Are these new people?
df.loc[df['customer_seniority_month'].isnull(), 'customer_registered_in_last6month'].unique()

array([1.])

As these are new customers, we give them the minimum seniority, i.e. 0.

In [38]:
df.loc[df['customer_seniority_month'].isnull(), 'customer_seniority_month'] = df['customer_seniority_month'].min()

In [39]:
df['customer_seniority_month'].isnull().sum()

0

### Date of Join

In [40]:
df['date_join'] = pd.to_datetime(df['date_join'])

In [41]:
df['date_join'].isnull().sum()

27734

Isn't it the same number that keeps repeating? These are new customers 

In [42]:
df.loc[df['date_join'].isnull(),'date_join'] = df['date_join'].median()

In [43]:
df['date_join'].isnull().sum()

0

### customer_loyalty
1 (First/Primary), 99 (Primary customer during the month but not at the end of the month)

In [44]:
df['customer_loyalty'].isnull().sum()

27734

In [45]:
df['customer_loyalty'].value_counts()

customer_loyalty
1.0     13594782
99.0       24793
Name: count, dtype: int64

In [46]:
df.loc[df['customer_loyalty'].isnull(), 'customer_loyalty'] = 1

### Province data

In [47]:
# address_type
df['address_type'].isnull().sum(), df['address_type'].unique()

(27735, array([ 1., nan]))

In [48]:
# province_code
df['province_code'].isnull().sum(), df['province_code'].unique()

(93591,
 array([29., 13., 50., 45., 24., 20., 10., 17., 49.,  8., 37.,  9., 22.,
        31.,  5., 40., 27., 25., 28.,  3., 42., 41., 39.,  7., 47., 36.,
        46., 44., 15., 32., 23., 16., 48., 12., 26.,  2.,  6., 30., 11.,
        nan,  4., 19., 34., 35., 14., 21., 18., 33., 38., 52., 43.,  1.,
        51.]))

Because the province name is given, we can drop these two columns

In [49]:
df.drop(['address_type', 'province_code'], axis=1, inplace=True)

### Is the client still active?

In [50]:
df['activity_index'].unique()

array([ 1.,  0., nan])

In [51]:
df['activity_index'].isnull().sum()

27734

In [52]:
df['activity_index'].value_counts()

activity_index
0.0    7384390
1.0    6235185
Name: count, dtype: int64

In [53]:
df.loc[df['activity_index'].isnull(), 'date_identifier'].sort_values()

261       2015-01-28
442435    2015-01-28
442399    2015-01-28
441951    2015-01-28
441833    2015-01-28
             ...    
3544890   2015-06-28
3544867   2015-06-28
3544790   2015-06-28
3547105   2015-06-28
3776332   2015-06-28
Name: date_identifier, Length: 27734, dtype: datetime64[ns]

it seems the most of the customers with null values in their activity index have had products in the first 6 months of the dataset. Who are these customers?

In [54]:
customeer_code = np.sort(df.loc[df['activity_index'].isnull(),'customer_code'].unique())


In [55]:
len(customeer_code)

7340

This is the number of customer codes (unique) where the activity index is null. Let's find if there are IDs from this list that they have not null values. 

In [56]:
some_cust_notnull_index = df.loc[(df.customer_code.isin(customeer_code)) & (df['activity_index'].notnull())].sort_values(by='date_identifier', 
                             ascending=False).groupby(['customer_code']).first()[['date_identifier', 'activity_index']]

In [57]:
some_cust_notnull_index.shape[0]

309

In [58]:
some_cust_notnull_index.head()

date_identifier  activity_index
customer_code                                
35152              2016-05-28             1.0
41449              2016-05-28             1.0
48623              2016-05-28             1.0
53148              2016-05-28             1.0
58520              2016-05-28             1.0

only 309 IDs have not null values. Lets match their most recent activity index found above for their null values. 

In [59]:
dff = df.copy()

In [60]:
dff['activity_index'] = dff['customer_code'].map(some_cust_notnull_index['activity_index']).fillna(dff['activity_index']) 

In [61]:
np.sort(dff.loc[dff['activity_index'].isnull(),'customer_code'].unique()).shape[0]

7031

Now we see that 309 customer codes are less! For the rest of customers, we don't have any clues, so we replace those with the median of activity index.

In [62]:
dff.loc[dff.activity_index.isnull(),"activity_index"] = dff["activity_index"].median()

In [63]:
dff['activity_index'].isnull().sum()

0

### Province name

There are some misspeling as well as null values.

In [64]:
dff.province_name.isnull().sum()

93591

In [65]:
dff.province_name.unique()

array(['MALAGA', 'CIUDAD REAL', 'ZARAGOZA', 'TOLEDO', 'LEON', 'GIPUZKOA',
       'CACERES', 'GIRONA', 'ZAMORA', 'BARCELONA', 'SALAMANCA', 'BURGOS',
       'HUESCA', 'NAVARRA', 'AVILA', 'SEGOVIA', 'LUGO', 'LERIDA',
       'MADRID', 'ALICANTE', 'SORIA', 'SEVILLA', 'CANTABRIA',
       'BALEARS, ILLES', 'VALLADOLID', 'PONTEVEDRA', 'VALENCIA', 'TERUEL',
       'CORUÑA, A', 'OURENSE', 'JAEN', 'CUENCA', 'BIZKAIA', 'CASTELLON',
       'RIOJA, LA', 'ALBACETE', 'BADAJOZ', 'MURCIA', 'CADIZ', nan,
       'ALMERIA', 'GUADALAJARA', 'PALENCIA', 'PALMAS, LAS', 'CORDOBA',
       'HUELVA', 'GRANADA', 'ASTURIAS', 'SANTA CRUZ DE TENERIFE',
       'MELILLA', 'TARRAGONA', 'ALAVA', 'CEUTA'], dtype=object)

In [66]:
dff.loc[dff.province_name=="CORU\xc3\x91A, A","province_name"] = "CORUNA, A"

For null values we create the following category:

In [67]:
dff.loc[dff.province_name.isnull(),"province_name"] = "UNKNOWN"

### gross_income_household

In [68]:
dff['gross_income_household'].isnull().sum()

2794375

Let's regionize these income values as replacing the null values with median income of all provinces is not good.

In [69]:
median_income_province= dff.groupby('province_name').agg(median_income=('gross_income_household',np.median)).reset_index()

C:\Users\srafi\AppData\Local\Temp\ipykernel_16272\3520925909.py:1: FutureWarning: The provided callable <function median at 0x00000221B64A56C0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  median_income_province= dff.groupby('province_name').agg(median_income=('gross_income_household',np.median)).reset_index()


In [70]:
median_income_province.head()

province_name  median_income
0         ALAVA       96967.65
1      ALBACETE       78657.30
2      ALICANTE       67526.28
3       ALMERIA       73390.65
4      ASTURIAS       87199.38

In [71]:
# replacing null values of a region with the corresponsing not null values in the same region
df_null_income = dff.loc[dff['gross_income_household'].isnull(),:]
merged_df = pd.merge(df_null_income, median_income_province, on='province_name', how='inner')

In [72]:
merged_df.gross_income_household.isnull().sum()

2794375

In [73]:
merged_df.gross_income_household = merged_df.median_income

In [74]:
merged_df.gross_income_household.isnull().sum()

0

now let's replace the null values in the main dff with these values

In [75]:
dff['gross_income_household'] = dff['customer_code'].map(merged_df['gross_income_household']).fillna(dff['gross_income_household'])

In [76]:
dff.gross_income_household.isnull().sum()

0

### Spouse_index

Let's see null values by now. from the following seris, spouse_index, and last_date_asprimary_customer_ifnot_endofmonth, so let's get rid of these two columns.

In [77]:
dff.isnull().sum().sort_values(ascending=False) * 100 / dff.shape[0]

spouse_index                                     99.986752
last_date_asprimary_customer_ifnot_endofmonth    99.818330
customer_segmentation_vip_indiv_colleggrad        1.387585
channel_used_tojoin                               1.363829
customer_relationtype_atbeginningofmonth          1.097513
customer_type_atbeginningofmonth                  1.097513
sex                                               0.203732
customer_dceased                                  0.203220
customer_residence_index                          0.203220
customer_foreigner_index                          0.203220
country_resid                                     0.203220
employee_index                                    0.203220
payroll                                           0.117701
pensions                                          0.117701
e_account                                         0.000000
longterm_deposits                                 0.000000
funds                                             0.0000

In [78]:
df.loc[df['last_date_asprimary_customer_ifnot_endofmonth'].notnull(),'customer_code']

509         1050477
2061        1049899
2624        1056267
4306        1054640
5451        1043896
             ...   
13642560    1173310
13642579    1173326
13645001    1164053
13646037    1168303
13646286    1168041
Name: customer_code, Length: 24793, dtype: int64

In [79]:
dff.drop(['spouse_index', 'last_date_asprimary_customer_ifnot_endofmonth'], axis=1, inplace=True)

In [81]:
dff.isnull().sum().sort_values(ascending=False)

customer_segmentation_vip_indiv_colleggrad    189368
channel_used_tojoin                           186126
customer_type_atbeginningofmonth              149781
customer_relationtype_atbeginningofmonth      149781
sex                                            27804
customer_foreigner_index                       27734
employee_index                                 27734
country_resid                                  27734
customer_dceased                               27734
customer_residence_index                       27734
pensions                                       16063
payroll                                        16063
mortgage                                           0
e_account                                          0
funds                                              0
longterm_deposits                                  0
mediumterm_deposits                                0
pensions_plan                                      0
date_identifier                               

### Payroll 

In [82]:
dff.payroll.isnull().sum()

16063

In [83]:
dff.payroll.unique()

array([ 0.,  1., nan])

In [84]:
dff.payroll.value_counts()

payroll
0.0    12885285
1.0      745961
Name: count, dtype: int64

Let's replace the nulls with 0s as it's dominant category

In [85]:
dff.loc[dff.payroll.isnull(),'payroll'] = 0

In [86]:
dff.payroll.isnull().sum()

0

### Pensions

In [87]:
dff.pensions.isnull().sum()

16063

In [88]:
dff.pensions.value_counts()

pensions
0.0    12821161
1.0      810085
Name: count, dtype: int64

In [89]:
dff.loc[dff.pensions.isnull(),'pensions'] = 0

### Sex

In [90]:
dff.sex.unique()

array(['H', 'V', nan], dtype=object)

In [91]:
dff.sex.isnull().sum()

27804

In [92]:
null_sex_cust_code = dff.loc[dff['sex'].isnull(), 'customer_code'].unique()

In [93]:
null_sex_cust_code, null_sex_cust_code.shape[0]

(array([1050741, 1051017, 1051064, ...,  429713,  278257,  216507],
       dtype=int64),
 7345)

In [94]:
cust_code_matching_notnull_cust = dff.loc[(dff['sex'].notnull())& (df['customer_code'].isin(null_sex_cust_code)), ['customer_code', 'sex']].groupby('customer_code').first()

In [95]:
cust_code_matching_notnull_cust.head()

sex
customer_code    
35152           V
41449           V
48623           V
53148           V
58520           V

In [96]:
dff['sex'] = dff['customer_code'].map(cust_code_matching_notnull_cust['sex']).fillna(dff['sex']) 


In [97]:
dff.sex.isnull().sum()

26755

In [98]:
dff["sex"].value_counts()

sex
V    7424931
H    6195623
Name: count, dtype: int64

In [99]:
dff.loc[dff.sex.isnull(),"sex"] = 'Unknown'

### Other columns with missing values

In [100]:
object_columns = dff.select_dtypes(include=["object"])
missing_columns = [col for col in object_columns if object_columns[col].isnull().any()]
for col in missing_columns:
    print("Unique values for {0}:\n{1}\n".format(col,object_columns[col].unique()))
del object_columns

Unique values for employee_index:
['N' nan 'A' 'B' 'F' 'S']

Unique values for country_resid:
['ES' nan 'CA' 'CH' 'CL' 'IE' 'AT' 'NL' 'FR' 'GB' 'DE' 'DO' 'BE' 'AR' 'VE'
 'US' 'MX' 'BR' 'IT' 'EC' 'PE' 'CO' 'HN' 'FI' 'SE' 'AL' 'PT' 'MZ' 'CN'
 'TW' 'PL' 'IN' 'CR' 'NI' 'HK' 'AD' 'CZ' 'AE' 'MA' 'GR' 'PR' 'RO' 'IL'
 'RU' 'GT' 'GA' 'NO' 'SN' 'MR' 'UA' 'BG' 'PY' 'EE' 'SV' 'ET' 'CM' 'SA'
 'CI' 'QA' 'LU' 'PA' 'BA' 'BO' 'AU' 'BY' 'KE' 'SG' 'HR' 'MD' 'SK' 'TR'
 'AO' 'CU' 'GQ' 'EG' 'ZA' 'DK' 'UY' 'GE' 'TH' 'DZ' 'LB' 'JP' 'NG' 'PK'
 'TN' 'TG' 'KR' 'GH' 'RS' 'VN' 'PH' 'KW' 'NZ' 'MM' 'KH' 'GI' 'SL' 'GN'
 'GW' 'OM' 'CG' 'LV' 'LT' 'ML' 'MK' 'HU' 'IS' 'LY' 'CF' 'GM' 'KZ' 'CD'
 'BZ' 'ZW' 'DJ' 'JM' 'BM' 'MT']

Unique values for customer_type_atbeginningofmonth:
[1.0 nan 3.0 2.0 '1.0' '1' '3' '3.0' '2.0' '4.0' 'P' '4' 4.0 '2']

Unique values for customer_relationtype_atbeginningofmonth:
['A' 'I' nan 'P' 'R' 'N']

Unique values for customer_residence_index:
['S' nan 'N']

Unique values for customer_foreigner

In [101]:
dff.loc[dff.customer_dceased.isnull(),"customer_dceased"] = "N"
dff.loc[dff.customer_relationtype_atbeginningofmonth.isnull(),"customer_relationtype_atbeginningofmonth"] = "A"
dff.customer_relationtype_atbeginningofmonth = dff.customer_relationtype_atbeginningofmonth.astype("category")

# As suggested by @StephenSmith
map_dict = { 1.0  : "1",
            "1.0" : "1",
            "1"   : "1",
            "3.0" : "3",
            "P"   : "P",
            3.0   : "3",
            2.0   : "2",
            "3"   : "3",
            "2.0" : "2",
            "4.0" : "4",
            "4"   : "4",
            "2"   : "2"}

dff.customer_type_atbeginningofmonth.fillna("P",inplace=True)
dff.customer_type_atbeginningofmonth = dff.customer_type_atbeginningofmonth.apply(lambda x: map_dict.get(x,x))
dff.customer_type_atbeginningofmonth = dff.customer_type_atbeginningofmonth.astype("category")

C:\Users\srafi\AppData\Local\Temp\ipykernel_16272\3884304447.py:19: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  dff.customer_type_atbeginningofmonth.fillna("P",inplace=True)


In [102]:
unknown_cols = [col for col in missing_columns if col not in ["customer_dceased","customer_relationtype_atbeginningofmonth","customer_type_atbeginningofmonth"]]
for col in unknown_cols:
    dff.loc[df[col].isnull(),col] = "UNKNOWN"

In [104]:
dff.isnull().sum()

date_identifier                               0
customer_code                                 0
employee_index                                0
country_resid                                 0
sex                                           0
age                                           0
date_join                                     0
customer_registered_in_last6month             0
customer_seniority_month                      0
customer_loyalty                              0
customer_type_atbeginningofmonth              0
customer_relationtype_atbeginningofmonth      0
customer_residence_index                      0
customer_foreigner_index                      0
channel_used_tojoin                           0
customer_dceased                              0
province_name                                 0
activity_index                                0
gross_income_household                        0
customer_segmentation_vip_indiv_colleggrad    0
saving_account                          

In [111]:
features

Column Name  \
0              fecha_dato   
1                ncodpers   
2            ind_empleado   
3         pais_residencia   
4                    sexo   
5                     age   
6              fecha_alta   
7               ind_nuevo   
8              antiguedad   
9                  indrel   
10         ult_fec_cli_1t   
11            indrel_1mes   
12            tiprel_1mes   
13                indresi   
14                 indext   
15               conyuemp   
16          canal_entrada   
17                indfall   
18                tipodom   
19               cod_prov   
20                nomprov   
21  ind_actividad_cliente   
22                  renta   
23               segmento   
24      ind_ahor_fin_ult1   
25      ind_aval_fin_ult1   
26       ind_cco_fin_ult1   
27      ind_cder_fin_ult1   
28       ind_cno_fin_ult1   
29      ind_ctju_fin_ult1   
30      ind_ctma_fin_ult1   
31      ind_ctop_fin_ult1   
32      ind_ctpp_fin_ult1   
33      ind_deco_fin_ult1   
34      ind_deme_fin_ult1   
35      ind_dela_fin_ult1   
36      ind_ecue_fin_ult1   
37      ind_fond_fin_ult1   
38       ind_hip_fin_ult1   
39      ind_plan_fin_ult1   
40      ind_pres_fin_ult1   
41      ind_reca_fin_ult1   
42      ind_tjcr_fin_ult1   
43      ind_valo_fin_ult1   
44       ind_viv_fin_ult1   
45        ind_nomina_ult1   
46      ind_nom_pens_ult1   
47        ind_recibo_ult1   

                                                                                                                                    Description  \
0                                                                                                      The table is partitioned for this column   
1                                                                                                                                 Customer code   
2                                                                   Employee index: A active, B ex employed, F filial, N not employee, P pasive   
3                                                                                                                  Customer's Country residence   
4                                                                                                                                Customer's sex   
5                                                                                                                                           Age   
6                                                           The date in which the customer became as the first holder of a contract in the bank   
7                                                                        New customer Index. 1 if the customer registered in the last 6 months.   
8                                                                                                                Customer seniority (in months)   
9                                                     1 (First/Primary), 99 (Primary customer during the month but not at the end of the month)   
10                                                                          Last date as primary customer (if he isn't at the end of the month)   
11  Customer type at the beginning of the month ,1 (First/Primary customer), 2 (co-owner ),P (Potential),3 (former primary), 4(former co-owner)   
12                            Customer relation type at the beginning of the month, A (active), I (inactive), P (former customer),R (Potential)   
13                                               Residence index (S (Yes) or N (No) if the residence country is the same than the bank country)   
14                                       Foreigner index (S (Yes) or N (No) if the customer's birth country is different than the bank country)   
15                                                                                     Spouse index. 1 if the customer is spouse of an employee   
16                                                                                           

In [110]:
dff.iloc[:,20:].sum(axis=1).head()

0    2.0
1    2.0
2    2.0
3    2.0
4    2.0
dtype: float64